openvdb/points/IndexFilter.h

```diff
@@ -142,15 +200,23 @@ class MultiGroupFilter
 
     inline bool initialized() const { return mInitialized; }
 
+    inline index::State state() const
+    {
+        return (mInclude.empty() && mExclude.empty()) ? index::ALL : index::PARTIAL;
```

Comment:
I think it's weird that this filter accepts no include groups as valid. imo it should be:

mInclude.empty() && mExclude.empty() = None
!mInclude.empty() && !mExclude.empty() = Partial
!mInclude.empty() && mExclude.empty() = Partial
mInclude.empty() && !mExclude.empty() = Partial

If the descriptor is known then you can optimise even further. This is a potentially destructive change, though I think it makes far more sense. Here is some custom logic I'm using to improve this behaviour. Assuming that a group does not exist in both the include and exclude groups, then:

```
        // The following assumes that a group does not exist inside both
        // the includeGroups and excludeGroups containers

        const bool hasExcludes(!excludeGroups.empty());
        for (auto iter = excludeGroups.begin(); iter != excludeGroups.end();) {
            if (!descriptor->hasGroup(*iter)) {
               std::swap(*iter, excludeGroups.back());
               excludeGroups.pop_back();
            }
            else {
                ++iter;
            }
        }

        // if exclude groups were specified but are now empty,
        // regardless of which include groups were specified we
        // process all points

        if (hasExcludes && excludeGroups.empty()) {
            includeGroups.clear();
            return ALL;
        }

        const bool hasIncludes(!includeGroups.empty());
        for (auto iter = includeGroups.begin(); iter != includeGroups.end();) {
            if (!descriptor->hasGroup(*iter)) {
               std::swap(*iter, includeGroups.back());
               includeGroups.pop_back();
            }
            else {
                ++iter;
            }
        }

        // if there were specified include groups but have all been removed
        // as they don't exist, check if any exclude groups exist. If not then
        // we shouldn't perform any execution

        if (hasIncludes && includeGroups.empty()) {
            if (!excludeGroups.empty()) return EXCLUDE;
            else return INVALID;
        }
```

Reply:
An empty includeGroups vector is simply treated as including "*" for convenience.

As for your logic here, this looks great. I consider this a further optimisation though, index::PARTIAL does produce the correct result, it's just not as efficient as it could be, as you highlight. There are other potential optimisations I didn't include too - BBoxFilter for example. Provided the results are correct and it doesn't introduce a regression, this can be introduced later I feel, or of course I'm more than happy if you'd like to add this to my PR with all the relevant unit tests. :)

Reply:
This isn't just an optimisation, it would change how the Houdini SOPs parse groups + anything which tries to use this filter with groups (names) which do no exist. Currently unavailable groups are dropped by the namesToIndices() convenience method indirectly through the first MultiGroupFilter constructor. This means that groups which don't exist are not handled "correctly". With the above:

- If exclude groups have been provided and there are no valid exclude groups, all points are considered regardless of include groups
- If include groups have been provided and there are no valid include groups, no points are considered regardless of exclude groups
- If include groups have been provided and there are no valid include groups BUT there are valid exclude groups, only the exclude groups are considered. This is not inversely true
- Otherwise use both include and exclude groups

Currently:

- If exclude groups have been provided and there are no valid exclude groups, the include groups are used (if these are also invalid/are not provided, everything is parsed)
- If include groups have been provided and there are no valid include groups, the exclude groups are used (if these are also invalid/are not provided, everything is parsed)

i.e: 

"!a b" - if a doesn't exist, everything should be used. Currently, only b will be used
"a b"  - if a and b don't exist, nothing should be used. Currently, everything will be used. Note this is the same for 'if "a" doesn't exist'. You should be using a GroupFilter for this, but a lot of algorithms don't.

You can try this stuff out with the current implementation from the SOPs. I hope I've explained this correctly. I've mentioned this before, (I think ^ behaviour also wrong but that's Houdini specific so slightly irrelevant) but as the filters are to be used directly and a lot of this is changing, it would be worthwhile to consider changing this.

Reply:
Hi @Idclip, thanks, your explanation makes perfect sense. The idea behind the state() method returning index::PARTIAL is that in this case, you cannot *guarantee* that all or no points will be evaluated. As such this maintains (status quo) correctness and performs *no slower* than the current mechanism. 

What you're talking about here is a worthwhile change, but I suggest could be tackled completely tangentially in a new PR (by modifying namesToIndices) to correct the behaviour and *then* introducing extra optimisations in the state() method once this extra optimisation framework in this PR has been merged in. Until then it just relies on returning index::PARTIAL which ultimately falls back to the implementation of the valid() index filter method. How does that sound?

Reply:
I appreciate what you're saying and I think the most prominent point to consider here is regarding backwards-compatibility. If we were to introduce behaviour-changing logic to the MultiGroupFilter, the only way to effectively retain backwards compatibility would be to have two MultiGroupFilter implementations, keep the old one for the deprecated includeGroups/excludeGroups implementation and introduce a new one when used with the explicit MultiGroupFilter parameter method. Essentially, if we were to rigorously uphold the existing behaviour, we would also then be forced to branch the SOPs that use it so that a newly-placed SOP would get the new behaviour and an existing one keep the old behaviour. This is a lot of work and perhaps more an argument to keep the existing (incorrect) behaviour. 

Ultimately, for the relatively minor use cases (in my opinion) that your changes will tidy up, I'd still rather that we include behaviour-changing logic as a separate PR, because there's also an impact in delaying an imminent change to the API.